Update libcoral for Python 3.11 and modern versions of Tensorflow

feranick opened the pull request:
libcoral (and pycoral that depends on it, making both the google-recommended ways to interact with EdgeTPU with C++ and Python respectively) are outdated due to several deprecations and changes within TF. This PR aims at updating libcoral by fixing such deprecations, modernizing the bazel building infrastructure, and docker files. 

One change is needed in TF to allow compilation. A separate issue is on file with separate PR [Change visibility status for lite/schema/schema_conversion_utils to allow compilation of external libraries](https://github.com/tensorflow/tensorflow/issues/63074).

There may be more issue to fix, but this PR provides a significant attempt at modernize the codebase. 

Reference issue: https://github.com/google-coral/pycoral/issues/137

feranick committed:
Synced with latest commit of libedgetpu

libedgetpu has been updated to support TF 2.15.0.

feranick committed:
Remove outdated deb package for python 2 in docker.

feranick committed:
Update version fo bazel to 6.1.0 in docker needed for TF2.15.0

feranick committed:
Update WROKSPACE to allow compilatin of current version of TF

1. Reworked after deprecation fo Tensorflow/toolchains.
2. Added definition of TF 2.15.0
3. Updated google benchmark to fix missing header call.

feranick committed:
Fix FTFS due to change in Eigen3 API

Replace deprecated Eigen::all with Eigen::indexing::all

feranick committed:
Replace signature_def_names with signature_keys

This is for tflite::Interpreter
The former call is deprecated, replaced by the latter

feranick committed:
Fix ambiguous call leading to FTFS, by overriding TfLiteFloatArrayCopy

feranick committed:
Update Dockerfile.windows

feranick committed:
Updated README.md

To include modern OS /Python versions

feranick committed:
Fix additional ambiguous call leading to FTFS, by overriding TfLiteFloatArrayCopy

feranick committed:
Fix additional FTFS due to change in Eigen3 API
Replace deprecated Eigen::last with Eigen::placeholders::all

feranick commented:
cc: @namburger @pkgoogle

feranick committed:
Use std=c++17 not c++14

Several compilation errors are fixed

feranick commented:
@namburger @pkgoogle, I can confirm that with this PR (and the manual application of https://github.com/tensorflow/tensorflow/issues/63074, compilation of libcoral is successful in both MacOS and Linux. As far as I can see this is ready to be merged.

Namburger commented:
Thanks @feranick! I don't think it's a good idea to merge this until we have https://github.com/tensorflow/tensorflow/issues/63074 figured out. I prefer not having users checking out project and hacking cached buildtool configs to get things to build. Although I can see the significant impact of having this repo updates and getting up to date with tensorflow. I'll try to see if we can gain more visibility for https://github.com/tensorflow/tensorflow/pull/63075

feranick commented:
Absolutely agree. In fact, even if https://github.com/tensorflow/tensorflow/pull/63075 and https://github.com/tensorflow/tensorflow/pull/63074 are out, they will certainly be on TF master, and hopefully in time for TF 2.16.0 but certainly not for 2.15.0 (unless it's backported on future maintenance release). So I would think this could be merged after TF 2.16.0 is out possibly with the fix above...

feranick committed:
Fix position of comment on TF version in WORKSPACE

feranick committed:
Updated rules-python to v0.31.0 and simplification of WORKSPACE

With rules-python v.0.27.0 intruduced a better way to handle pypi requirements, removing the need of
 several calls within WORKSPACE.

feranick committed:
Temporarily use fork of libedgetpu with initial support for TF2.16.

Stable libedgetpu only supports TF2.15.0. It will be reverted back to the official repo once that has been updated to TF2.16.0

feranick committed:
Sync with latest feranick/libedgetpu

feranick committed:
Add support for TF 2.16.0-rc0

feranick committed:
Sync with latest libedgetpu commit

feranick committed:
Revert commit aae943b5eca

While the original commit simplified WORKSPACE, it is not alligned with the current version of TF. REverting to the previous version of rules_python v.0.26.0 which is the same shipped with TF 2.16.0.

feranick committed:
Sync with libedgetpu

feranick committed:
Use more modern version fo debian in docker.mk

feranick committed:
Update distributions in build script.
Additional calls to gstreamer-1.-1

feranick committed:
Sync with latest libedgetpu

feranick committed:
Build against TF2.17.0-dev to resolve visibility issue in SCHEMA

A visibility issue in TF SCHEMA was preventing compilation of libcoral.
Ref: https://github.com/tensorflow/tensorflow/issues/63074

This is now fixed in TF 2.17.0-dev with commit 79ecb3f

feranick commented:
Compilation of libcoral works fine with MacOS and Linux by building against TF 2.17.0-dev commit 79ecb3f, where the [visibility issue](https://github.com/tensorflow/tensorflow/commit/79ecb3f8bb6bd73f0115fa9a97b630a6f745a426) is now fixed.

feranick commented:
@Namburger I already found a possible breakage in compilation for `arm` due to 2.17 [(See log here)](https://github.com/google-coral/libcoral/files/14501456/log_arm.txt) when crosscompiling for arm (there are no issues in MacOS and Linux for x86). I forked 2.16.0 and added the visibility patch to test whether the issue is in 2.17 or not. 

When compiling against a local fork of TF r2.16 with the visibility patch, the above issue is not present, but another one is specific to libcoral and arm. Log attached.
[log_arm_TF2.16.0.txt](https://github.com/google-coral/libcoral/files/14501479/log_arm_TF2.16.0.txt)

Namburger commented:
> @Namburger I already found a possible breakage in compilation for `arm` due to 2.17 [(See log here)](https://github.com/google-coral/libcoral/files/14501456/log_arm.txt) when crosscompiling for arm (there are no issues in MacOS and Linux for x86). I forked 2.16.0 and added the visibility patch to test whether the issue is in 2.17 or not.
> 
> When compiling against a local fork of TF r2.16 with the visibility patch, the above issue is not present, but another one is specific to libcoral and arm. Log attached. [log_arm_TF2.16.0.txt](https://github.com/google-coral/libcoral/files/14501479/log_arm_TF2.16.0.txt)

@Namburger I already found a possible breakage in compilation for `arm` due to 2.17 [(See log here)](https://github.com/google-coral/libcoral/files/14501456/log_arm.txt) when crosscompiling for arm (there are no issues in MacOS and Linux for x86). I forked 2.16.0 and added the visibility patch to test whether the issue is in 2.17 or not.

When compiling against a local fork of TF r2.16 with the visibility patch, the above issue is not present, but another one is specific to libcoral and arm. Log attached. [log_arm_TF2.16.0.txt](https://github.com/google-coral/libcoral/files/14501479/log_arm_TF2.16.0.txt)

feranick committed:
Add arm-specific compiler flags to fix compilation due to undedefined variable types.

feranick commented:
I fixed the compilation with TF 2.16.0 and visibility patch by adding --cxxopt=-mfp16-format=ieee to arm specific flags. 

Compilation is successful now across Linux architectures. Will go back and see what is going on with 2.17.0.

feranick committed:
Revert previous commit for aarch64

feranick committed:
Revert commit 139442b also for armhf

Not needed for clean build with Docker.

feranick committed:
Use bazel 6.5.0 for TF 2.17.0

feranick committed:
Sync with libedgetpu

feranick commented:
@Namburger Since this PR seems to work for the most part (i.e. MacOS, Linux_X86-64, Linux_aarch64), I was wondering if would make sense to merge now that it builds to a workable version of TF 2.17.0. It will be in a better state than it currently is (where it doesn't build anywhere). A couple of observations:
1. If you think we can go ahead, we will need to do it in stages, i.e. first merging an updated version of `libedgetpu`, then `libcoral`, then `pycoral` all built against the same TF 2.17.0 with visibility patch. Once each has been merged, the next needs to be quickly adapted to the right repo/commit, which I can do.
2. `libedgetpu` builds and works fine with TF 2.16.0, but it will need to be built against TF2.17.0 to be compatible with libcoral/pycoral. Thoughts?

Of course the other option is to wait until much later when 2.17.0 is ready, but that would mean not having pycoral at all for quite a while. Please let me know.

Namburger commented:
> @Namburger Since this PR seems to work for the most part (i.e. MacOS, Linux_X86-64, Linux_aarch64), I was wondering if would make sense to merge now that it builds to a workable version of TF 2.17.0. It will be in a better state than it currently is (where it doesn't build anywhere). A couple of observations:
> 
> 1. If you think we can go ahead, we will need to do it in stages, i.e. first merging an updated version of `libedgetpu`, then `libcoral`, then `pycoral` all built against the same TF 2.17.0 with visibility patch. Once each has been merged, the next needs to be quickly adapted to the right repo/commit, which I can do.
> 2. `libedgetpu` builds and works fine with TF 2.16.0, but it will need to be built against TF2.17.0 to be compatible with libcoral/pycoral. Thoughts?
> 
> Of course the other option is to wait until much later when 2.17.0 is ready, but that would mean not having pycoral at all for quite a while. Please let me know.

@feranick appreciate all your work so far, I'm trying to work it out, it's not promising that we keep running into one issue after another.. I don't know if it's a good idea to break things for armv7 forks at the moment. I'm talking to some forks who pushed the change that broke the build will try to get thing working..

Full disclaimer I'm currently having other priories and coral work is pure volunteering, so I can only do so much. Other forks I'm talking to are also working on other things so I can't push too hard, but I understand the importance of keeping this project updated and will try my best to keep things moving asap. Please see https://github.com/google-coral/libedgetpu/pull/60#issuecomment-1981154554

feranick committed:
Reinstate revised version of commit 139442bf1

feranick commented:
@namburger, I absolutely understand, your interest even voluntary is great and much appreciated. Let's see what happens, and if we can find ways to push things a bit at the time, that is great anyway. I'll be here for any help I can provide.

Thanks!

feranick committed:
Resync with libedgetpu

feranick committed:
Update stable TF version required to 2.16.1

This is disable in favor of the 2.17.0-dev with visibility patch.

Commit:
Update file for dh-python, needed to build debian packages

feranick committed:
Sync with libedge for TF 2.17.0-rc0

feranick committed:
Updated dockerfile.windows with latest dependencies

feranick committed:
Added support for temsorflow 2.17.0-rc1

feranick committed:
Add support for TF 2.17.0 stable

feranick commented:
Before this can be merged, a new PR for `libedgetpu` needs to be pushed first (as it adds support for TF 2.17.0).

https://github.com/google-coral/libedgetpu/pull/70

feranick committed:
Added support for Ubuntu:24.04

24.04 introduces a new way to handle repositories, using sources rather than lists.

feranick committed:
Change su user to root to allow compilation in Ubuntu 24.04

feranick committed:
Pass correct TF_PYTHON_VERSION during build.

feranick committed:
Updated third party libraries in Docker.windows

feranick committed:
Merge pull request #1 from tranzmatt/master

Update file for dh-python, needed to build debian packages

feranick committed:
Added support for TF 2.17.1

feranick committed:
Typo fixes

feranick committed:
Initial support for tensorflow 2.18.1

feranick committed:
Add explicit support for Apple silicon

Rather than referring to only "Darwin" use
darwin_arm64 darwin_x86_64

feranick committed:
Updated Dockerfile.windows

feranick committed:
Fix typo

feranick committed:
Added initial support for Tensorflow 2.19.1
